backend/src/main/java/com/soen341/instagram/service/impl/RegistrationService.java

```diff
@@ -0,0 +1,75 @@
+package com.soen341.instagram.service.impl;
+
+import java.util.Date;
+import java.util.regex.Pattern;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+
+import com.soen341.instagram.dao.impl.AccountRepository;
+import com.soen341.instagram.exception.account.EmailTakenException;
+import com.soen341.instagram.exception.account.InvalidEmailFormatException;
+import com.soen341.instagram.exception.account.InvalidUsernameFormatException;
+import com.soen341.instagram.exception.account.UsernameTakenException;
+import com.soen341.instagram.model.Account;
+
+@Service("registrationService")
+public class RegistrationService
+{
+	@Autowired
+	private AccountRepository accountRepository;
+
+	@Autowired
+	private PasswordEncoder passwordEncoder;
+
+	public void createNewAccount(final Account account)
+	{
+		checkIfEmailTaken(account.getEmail());
+
+		checkIfUsernameTaken(account.getUsername());
+
+		checkIfEmailFormatValid(account.getEmail());
+
+		checkIfUsernameFormatValid(account.getUsername());
+
+		account.setPassword(passwordEncoder.encode(account.getPassword()));
+		account.setCreated(new Date());
+		accountRepository.save(account);
+	}
+
+	private void checkIfEmailTaken(final String email)
+	{
+		if (!(accountRepository.findByEmail(email) == null))
+		{
+			throw new EmailTakenException();
+		}
+	}
+
+	private void checkIfUsernameTaken(final String username)
+	{
+		if (!(accountRepository.findByUsername(username) == null))
+		{
+			throw new UsernameTakenException();
+		}
+	}
+
+	private void checkIfEmailFormatValid(final String email)
+	{
+		String regex = "(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|\"(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21\\x23-\\x5b\\x5d-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])*\")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21-\\x5a\\x53-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])+)\\])";
+		if (!Pattern.matches(regex, email))
+		{
+			throw new InvalidEmailFormatException();
+		}
+	}
+
+	private void checkIfUsernameFormatValid(final String username)
```

Comment:
I was thinking that this code is actually useful in the profile service as well... Instead of duplication maybe we can create some sort of abstraction (an interface or something for checking username format, taken usernames, etc.) This does not have to be done immediately, maybe some later refactoring?

Reply:
Yes definitely agree, I thought of it too, but like you said it could be a refactored later. Maybe we could create some sort of class that handle these verification. Maybe you could create it for your tickets and refactored my code later?

Reply:
I will do that.